Persistent tower

ryoqun opened the pull request:
#### Problem

tower isn't persisted.

#### Summary of Changes

Persist it

#### Todo

- [x] Address any remaining review comments from #7436 #9902 

Fixes #6936

ryoqun commented on core/src/consensus.rs:
```diff
@@ -544,13 +603,11 @@ impl Tower {
         node_pubkey: &Pubkey,
     ) -> Option<Arc<Bank>> {
-        let ancestors = bank_forks.read().unwrap().ancestors();
+        let ancestors = bank_forks.ancestors();
```
Hmm, broader lock might affect performance?

ryoqun replied:
Well, this isn't issue because this is only called from the Tower restoration codepath.

ryoqun commented on sdk/src/slot_history.rs:
```diff
@@ -65,14 +65,18 @@ impl SlotHistory {
     pub fn check(&self, slot: Slot) -> Check {
         if slot >= self.next_slot {
             Check::Future
-        } else if self.next_slot - slot > MAX_ENTRIES {
```
Well, this causes panic if given slot is large enough. Separate pr....

ryoqun replied:
Well, I was lying... https://github.com/solana-labs/solana/pull/10718/files#diff-b6bc75c0bdcd2062ec00c3da3d974053R66 ensures that the overflow doesn't happen....

ryoqun commented on core/src/consensus.rs:
```diff
@@ -602,6 +659,87 @@ impl Tower {
+            .votes
+            .retain(move |_| is_diverged_descendants_iter.next().unwrap());
+
+        self.lockouts.root_slot = Some(replayed_root_slot);
```
hmm, I shouldn't do this here?

ryoqun replied:
Well, I'm now sure this is actually needed.

ryoqun commented on local-cluster/tests/local_cluster.rs:
```diff
@@ -1148,6 +1151,127 @@ fn test_no_voting() {
     }
 }
 
+#[test]
```
maybe add an test case for this? https://github.com/solana-labs/solana/pull/9902#issuecomment-627875209

ryoqun replied:
Done at test_switch_threshold_across_tower_reload:
https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR1323

ryoqun commented on core/src/consensus.rs:
```diff
@@ -641,6 +785,121 @@ impl Tower {
+            )));
+        }
+
+        fs::create_dir_all(&self.save_path)?;
```
I guess this check can be omitted from this rather critical codepath.

ryoqun commented on core/src/consensus.rs:
```diff
@@ -641,6 +785,121 @@ impl Tower {
+            let mut file = File::create(&new_filename)?;
+            let saveable_tower = SavedTower::new(self, node_keypair)?;
+            bincode::serialize_into(&mut file, &saveable_tower)?;
+            file.sync_all().unwrap();
```
Add metrics for this whole code block.

ryoqun replied:
done

ryoqun replied:
Well, this wasn't done until now. Not sure why I marked this resolved 6 days ago...

ryoqun commented on core/src/validator.rs:
```diff
@@ -562,10 +567,21 @@ impl Validator {
     }
 }
 
+fn empty_vote_account(bank: &Arc<Bank>, vote_account: &Pubkey) -> Option<bool> {
```
odd return type

ryoqun replied:
done!

ryoqun commented on core/src/heaviest_subtree_fork_choice.rs:
```diff
@@ -456,15 +456,20 @@ impl ForkChoice for HeaviestSubtreeForkChoice {
+                let heaviest_slot_on_same_voted_fork =
+                    self.best_slot(last_vote).unwrap_or_default();
+                if heaviest_slot_on_same_voted_fork == 0 {
+                    error!("fix me.....");
```
fix me...

ryoqun replied:
finally done!

ryoqun commented on core/src/consensus.rs:
```diff
@@ -648,6 +795,121 @@ impl Tower {
+            let mut file = File::create(&new_filename)?;
+            let saveable_tower = SavedTower::new(self, node_keypair)?;
+            bincode::serialize_into(&mut file, &saveable_tower)?;
+            //file.sync_all().unwrap();
```
Well, it turned out this and the other `.sync_all()` causes a lot of instability even with local cluster tests...

ryoqun commented on runtime/src/accounts_index.rs:
```diff
@@ -12,10 +12,11 @@ pub type Ancestors = HashMap<Slot, usize>;
 
 pub type RefCount = u64;
 type AccountMapEntry<T> = (AtomicU64, RwLock<SlotList<T>>);
+pub type AccountMap<K, V> = BTreeMap<K, V>;
```
rebase!

ryoqun commented on core/src/consensus.rs:
```diff
@@ -602,6 +663,99 @@ impl Tower {
+        replayed_root_slot: Slot,
+        slot_history: &SlotHistory,
+    ) -> Result<Self> {
+        error!("adjust_lockouts_after_replay....");
```
remove this

ryoqun commented on core/src/validator.rs:
```diff
@@ -626,6 +648,46 @@ fn new_banks_from_blockstore(
+        .and_then(|tower| {
+            let root_bank = bank_forks.root_bank();
+            let slot_history = root_bank.get_slot_history();
+            // assert bank is frozen
```
I think this is redundant; remove just this.

ryoqun commented on core/src/consensus.rs:
```diff
@@ -1788,4 +2241,335 @@ pub mod test {
+    }
+
+    #[test]
+    fn test_expire_old_votes_on_load2() {
```
more descriptive test case name, please.

ryoqun commented on core/src/consensus.rs:
```diff
@@ -1066,6 +1367,158 @@ pub mod test {
     }
 
+    #[test]
+    fn test_switch_threshold_across_tower_reload() {
```
Add comment at each steps...

ryoqun commented on validator/src/main.rs:
```diff
@@ -727,6 +727,12 @@ pub fn main() {
                 .help("Use CUDA"),
         )
+        .arg(
+            clap::Arg::with_name("require_tower")
```
`--require-tower` will soon be the default. :)

ryoqun commented on core/src/heaviest_subtree_fork_choice.rs:
```diff
@@ -480,6 +480,10 @@ impl ForkChoice for HeaviestSubtreeForkChoice {
         let heaviest_slot_on_same_voted_fork = last_voted_slot.map(|last_vote| {
+            if tower.is_restored_stray_slot(last_vote) {
+                // return None because bank_forks doesn't have corresponding banks
+                return None;
```
write a test for this.

ryoqun replied:
done!

ryoqun commented on core/src/consensus.rs:
```diff
@@ -400,9 +462,20 @@ impl Tower {
+                let last_vote_ancestors = ancestors.get(&last_voted_slot).unwrap_or_else(|| {
+                    if self.is_restored_stray_slot(last_voted_slot) {
+                        // Construct artifical ancestors because we can't with ancestors from bank_forks
+                        maybe_stray_ancestors = self.restored_stray_slots.range(0..last_voted_slot).copied().collect();
```
write a test for this, showing different behavior with previous implementation (returning HashSet::default()).

ryoqun replied:
done!

ryoqun commented on core/src/consensus.rs:
```diff
@@ -70,13 +79,20 @@ pub(crate) struct ComputedBankState {
+    #[serde(skip)]
+    tmp_path: PathBuf, // used before atomic fs::rename()
+    #[serde(skip)]
+    restored_stray_slots: BTreeSet<Slot>,
```
write in-source and on-pr comments for this, including its design reasonings.

ryoqun commented on core/src/validator.rs:
```diff
@@ -564,8 +569,61 @@ impl Validator {
+            tower.adjust_lockouts_after_replay(root_bank.slot(), &slot_history)
+        })
+        .unwrap_or_else(|err| {
+            let is_already_active = active_vote_account_exists_in_bank(&bank_forks.working_bank(), &vote_account);
```
Hmm, maybe should this be `bank_forks.root_bank()`? Otherwise, can any bad validator trick victim validator with crafted block with high-number crafted slot, somehow?

ryoqun commented on core/src/validator.rs:
```diff
@@ -627,6 +694,21 @@ fn new_banks_from_ledger(
+    );
+
+    info!(
+        "Tower state: root slot={:?}, last vote slot={:?}",
```
This is one-off loggging, so I'd rather like to print the whole thing to make debugging easier.

ryoqun commented on core/src/consensus.rs:
```diff
@@ -73,14 +83,20 @@ pub(crate) struct ComputedBankState {
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, AbiExample)]
 pub struct Tower {
     node_pubkey: Pubkey,
     threshold_depth: usize,
     threshold_size: f64,
```
I wonder these should be serialized, assuming these could be dynamic in the future?

carllin replied:
yeah so if these change in the future, we'll need some sort of rolling upgrade to make sure people all use the same values at the same time...could be tricky

ryoqun commented on core/src/consensus.rs:
```diff
@@ -73,14 +83,20 @@ pub(crate) struct ComputedBankState {
 }
 
 #[frozen_abi(digest = "2ZUeCLMVQxmHYbeqMH7M97ifVSKoVErGvRHzyxcQRjgU")]
```
Cool, #8012 ensures abi doesn't change even if we're adding new fields to this struct as long as they're `#[serde(skip)]`.

ryoqun commented on core/src/consensus.rs:
```diff
@@ -406,7 +468,16 @@ impl Tower {
         self.last_voted_slot()
             .map(|last_voted_slot| {
-                let last_vote_ancestors = ancestors.get(&last_voted_slot).unwrap();
+                let last_vote_ancestors = ancestors.get(&last_voted_slot).unwrap_or_else(|| {
```
Test that `self.is_stray_last_vote()` should be checked only if `ancestors.get() -> None`

ryoqun commented on core/src/heaviest_subtree_fork_choice.rs:
```diff
@@ -525,6 +525,26 @@ impl HeaviestSubtreeForkChoice {
 
+    fn heaviest_slot_on_same_voted_fork(&self, tower: &Tower) -> Option<Slot> {
+        tower.last_voted_slot().map(|last_voted_slot| {
+            let heaviest_slot_on_same_voted_fork = self.best_slot(last_voted_slot);
```
Add a test that `is_stray_last_vote()` should be checked only after `self.best_slot() -> None`...

ryoqun replied:
done

ryoqun commented on core/src/consensus.rs:
```diff
@@ -651,6 +913,137 @@ impl Tower {
+    InconsistentWithSlotHistory(String),
+}
+
+#[derive(Default, Clone, Serialize, Deserialize, Debug, PartialEq)]
```
Add frozen abi here too.

ryoqun commented on core/src/consensus.rs:
```diff
@@ -404,9 +478,28 @@ impl Tower {
+                        );
+                        &stray_restored_ancestors
+                    } else {
+                        panic!("no ancestor!")
```
this is same as the `.unwrap()` in the original code.

ryoqun replied:
now outdated comment...

ryoqun commented:
Wow, I needed unexpectedly large amount of corner case handling... I think I've finally managed to address any imaginable concerns in the implementation with enough comments.

Status: remaining is light on-PR commentary; after that, I think this finally can graduate from the draft!!!

ryoqun commented on core/src/consensus.rs:
```diff
@@ -73,14 +83,32 @@ pub(crate) struct ComputedBankState {
+    // validator's restart porpose.
+    // This could be emptied after some time; but left intact indefinitely for easier
+    // implementation
+    stray_restored_slots: BTreeSet<Slot>,
```
a new adjective `stray` is introduced to mean slots which aren't found in `bank_forks`.

ryoqun replied:
Being `BTreeSet` not `HashSet` is needed because of https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR716.

carllin replied:
github isn't showing me the diff in the link 😢

ryoqun commented on core/src/consensus.rs:
```diff
@@ -651,6 +945,144 @@ impl Tower {
+            let mut file = File::create(&new_filename)?;
+            let saved_tower = SavedTower::new(self, node_keypair)?;
+            bincode::serialize_into(&mut file, &saved_tower)?;
+            // file.sync_all() hurts performance; pipeline sync-ing and submitting votes to the cluster!
```
I'm aware of https://github.com/solana-labs/solana/pull/9902#discussion_r442439449, but I just noticed this introduces moderate latencies even with casual testing and local_cluster tests. So, I'm commenting out this for now.

I'll address the pipe-lining if this durability is really desired. Well, assuming most of validators are hosted at data-center, the power-outages and/or kernel-crashes should be very rare.

carllin replied:
yeah agreed. Out of curiosity, do you know about how long this `sync_all` was taking?

ryoqun replied:
@carllin Well, this depends on various factors including hardware, os block system config, filesystem, etc. On my machine, enabling this easily caused 1~2s spikes...

ryoqun commented on core/src/consensus.rs:
```diff
@@ -404,9 +486,26 @@ impl Tower {
             .map(|last_voted_slot| {
-                let last_vote_ancestors = ancestors.get(&last_voted_slot).unwrap();
+                let last_vote_ancestors = if !self.use_stray_restored_ancestors(ancestors, last_voted_slot) {
+                    ancestors.get(&last_voted_slot).unwrap()
```
Relaxing an invariant of bank_forks by removing `unwrap()` here might hide other bad conditions. So, I retained `.unwrap()` for the general code path as opposed to original diff: https://github.com/solana-labs/solana/pull/9902/files#diff-d848cba681535238849e7e33ed32702bL353

ryoqun commented on core/src/consensus.rs:
```diff
@@ -395,6 +469,14 @@ impl Tower {
         false
     }
 
+    fn use_stray_restored_ancestors(
```
Admittedly, I extracted this `fn` only for unit-tests: https://github.com/solana-labs/solana/pull/10718/commits/8c65c2ea59050a6992472c49b31d4c4f2fa677c0.

This code-arrangement itself incurs another needless lookup of `HashMap<Slot, HashSet<Slot>>` for each `check_switch_threshold` invocation. But I think this is tolerable.

carllin replied:
I see your point,  because when this is called in `check_switch_threshold` we already know, `ancestors.contains_key(&slot)`. Since this first check is only really used in tests, can we make this a test only function and replace the call to `use_stray_restored_ancestors()` with just `self.is_stray_last_vote()` below?

carllin replied:
I see your point,  because when this is called in `check_switch_threshold` we already know `ancestors.contains_key(&slot)`. Since this first check is only really used in tests, can we make this a test only function and replace the call to `use_stray_restored_ancestors()` with just `self.is_stray_last_vote()` (more clarity) below?

ryoqun replied:
done!

ryoqun commented on core/src/consensus.rs:
```diff
@@ -606,6 +703,203 @@ impl Tower {
+            } else {
+                self.stray_restored_slots = self.voted_slots().into_iter().collect();
+            }
+            // should call self.votes.pop_expired_votes()?
```
This is the only remaining question in this pr.

ryoqun commented on core/src/consensus.rs:
```diff
@@ -606,6 +703,203 @@ impl Tower {
+
+        // iterate over votes in the newest => oldest order
+        // bail out early if bad condition is found
+        for voted_slot in checked_slots.iter().rev() {
```
This is the heart of [this logic](https://github.com/solana-labs/solana/pull/9902#issuecomment-627875209).

I've tried to do best to detect any abnormalities, while doing this with 1-pass. So, I've hand-rolled this rather complicated `while` looping construct.

ryoqun commented on core/src/consensus.rs:
```diff
@@ -606,6 +703,203 @@ impl Tower {
+        let mut root_checked = false;
+        let mut checked_slots = if let Some(root) = self.lockouts.root_slot {
+            root_checked = true;
+            vec![root]
```
Extending [this logic](https://github.com/solana-labs/solana/pull/9902#issuecomment-627875209), I needed to include root for this anchoring process, because if all slots in tower except the root was on minor fork, [this error check incorrectly triggers](https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR856-R857).

This might sound really edge-case? Unfortunately, this isn't. This always occurs with single-leader clusters such as `./run.sh`.

ryoqun commented on core/src/heaviest_subtree_fork_choice.rs:
```diff
@@ -498,6 +498,30 @@ impl HeaviestSubtreeForkChoice {
+                    // return None because bank_forks doesn't have corresponding banks
+                    return None;
+                };
+                let heaviest_slot_on_same_voted_fork = heaviest_slot_on_same_voted_fork.expect(
```
Similar to this [reasoning](https://github.com/solana-labs/solana/pull/10718/files#r455530911), I retained `.expect()` here. Btw, these code wasn't existing at #9902.

ryoqun commented on core/src/consensus.rs:
```diff
@@ -606,6 +703,203 @@ impl Tower {
+
+        // return immediately if votes are empty...
+        if self.lockouts.votes.is_empty() {
+            assert_eq!(self.root(), None);
```
I guess if votes are empty, root is always `None`.

ryoqun replied:
well, not anymore, I ended up being root is always Some.

ryoqun commented on core/src/consensus.rs:
```diff
@@ -651,6 +945,144 @@ impl Tower {
+    TooOldSlotHistory(Slot, Slot),
+
+    #[error("The tower is inconsistent with slot history: {0}")]
+    InconsistentWithSlotHistory(String),
```
catch-all-error category for really odd cases.

ryoqun commented on core/src/consensus.rs:
```diff
@@ -606,6 +703,203 @@ impl Tower {
+            self.voted_slots()
+        );
+
+        assert_eq!(slot_history.check(replayed_root_slot), Check::Found);
```
various assertions...

ryoqun commented on core/src/consensus.rs:
```diff
@@ -404,9 +486,26 @@ impl Tower {
+                } else {
+                    // Use stray restored ancestors because we couldn't derive them from
+                    // given ancestors (=bank_forks)
+                    // Also, can't just return empty ancestors because we should exclude
```
From https://github.com/solana-labs/solana/pull/9902#issuecomment-627875209,

> with `unwrap_or_default()`. With that change, these checks should all be skipped: 

I originally returned here `HashSet::default()` but it turned out this should return a set containing `G` (in the original term; in this pr, this is called `stray`).

The corresponding test assertion is https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR2404-R2405.

carllin replied:
hmmm so this is interesting. 

So first we note that the purpose of `last_vote_ancestors` is to make sure we don't count lockouts on ancestors of the last vote, i.e. to fulfill the check `!llast_vote_ancestors.contains(lockout_interval_start)`, where `lockout_interval_start` > root (checked at https://github.com/solana-labs/solana/blob/master/core/src/consensus.rs#L487)

Now in cases where blockstore is not corrupted, returning the default `HashSet::default()` here should be ok, because on restart from blockstore. To see this we consider the following two cases:

1) If the snapshot root is <= last voted slot, then: 
a) your local root must be a descendant/ancestor of the snapshot root (otherwise we reject that snapshot b/c no `anchor` slots will be found)
b) all your voted slots are descendants of the local root.

Then all the ancestors between the snapshot root and `last voted slot` are present in blockstore at startup and should be replayed at startup. This means `ancestors.get(&last_voted_slot).unwrap()` will return those ancestors. Now I think this test `test_switch_threshold_across_tower_reload()` is failing with `unwrap_or_default()` because when you simulate the reload, the `forks` tree is missing the votes in tower. I suspect if you change the forks tree here: https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR2437 to include the votes like so:

```
let forks = tr(0)
            / (tr(1)
                / (tr(2)
                    / tr(10)
                    / (tr(43)
                        / (tr(44)
                            // Minor fork 2
                            / (tr(45) / (tr(46) / (tr(47) / (tr(48) / (tr(49) / (tr(222)))))))
                            / (tr(110) / tr(111))))));
```

things should work. 

Now this does raise the question what happens if your blockstore is corrupted and you can't replay those voted slots on startup. I think the problem with the current approach is that the votes on the stack do not fully capture the ancestor history. For instance, if your voting history was `1, 2, 3, 6`, the stack will only show `1, 2, 6`. 

In such situations (I don't think we have to do these in this PR), I believe the correct thing to do would be a) Return `FailedSwitch` if `switch_slot < last_voted_slot`, and then if `switch_slot > last_voted_slot`, only count `lockout_interval_start > last_voted_slot` so you're sure `lockout_interval_start` is not an ancestor of `last_voted_slot`

2) If the snapshot root is > last voted slot, then because `lockout_interval_start > root` this means that that  `lockout_interval_slot` is not an ancestor of `last_voted_slot` means any  using `HashSet::default` will return the correct result for `!llast_vote_ancestors.contains(lockout_interval_start)`.

ryoqun replied:
>     1. otherwise we reject that snapshot b/c no `anchor` slots will be found)

Oops, we weren't rejecting that snapshot as a form of `TowerError` even if our local tower has been rooted. Fixed that: https://github.com/solana-labs/solana/pull/10718/commits/929ca6d859b697d54753e1e4b85c04c85d7822a0#r470455410

ryoqun replied:
Otherwise, your comment sounds valid. I'm still yet to fully grok this advanced consensus discussion, though ... ;)

ryoqun replied:
@carllin Hehe, quite dense discussion. My head is spinning hot!

Well, your thinking is correct under the assumption that _tower always has root unless something is corrupted_. But, I think otherwise: _restored tower can legitimately have **no root** if the tower is very new and warming up_. I encoded the case as code and test respectively: https://github.com/solana-labs/solana/pull/10718/files#r470502855 and https://github.com/solana-labs/solana/pull/10718/files#r470501988

So, borrowing your logic, here is quite corner of corner of **Case 1** (not your **Case 2**) I've tried to address with the current code xD:

if ...
- (i) the newly-born tower just unfortunately starts to vote on a minor fork for several slots,
- (ii) and the validator is immediately stopped before rooting the tower,
- (iii) and the validator is later restarted with a snapshot which is **after** (not **before**) the common ancestor of both the minor fork and the major (of snapshot) fork (this should create a situation where _there are votes which aren't descendents of local root after replay_),

,then

- (x) the voted slots should be unavailable in the bank_forks and therefore,
- (y) there need to be some way (= stray slots) to exclude `lockouts` on an ancestors of the last vote because other validator might have legitimately voted on the same minor fork as well. Note that those ancestors can be newer than the local root (= `lockout_interval_start > root`).

Am I right? Be aware of that I'm still not expert in this consensus area. Or the your mentionedloptimistic confirmation rule doesn't need to enforced in this case?  I might be blatantly wrong... ;)

carllin replied:
@ryoqun my head is spinning too :)

Let me see if I can properly illustrate the example you mentioned above:

```
      1----------8---------10 (last vote on minor fork)
   /
0
    \
       2 (snapshot) -- 3
```

1) The validator voted on `1,  8, 10`, didn't set a root, then crashed
2) The validator restarts from a snapshot of `2`

Now let's consider your concern: "there need to be some way (= stray slots) to exclude lockouts on an ancestors of the last vote because other validator might have legitimately voted on the same minor fork as well. Note that those ancestors can be newer than the local root". 

So in our example:

1) "those ancestors can be newer than the local root" - The "local root" here would be slot `2`, so an example off such an ancestor would be slot `8 > 2` in the above example.

2) "there need to be some way (= stray slots) to exclude lockouts on an ancestors of the last vote". Here, because slot `8` is not descended from the root `slot 2`, it won't exist in BankForks, so it's naturally excluded.

Does that make sense? Or did I misunderstand your concern 😄

ryoqun replied:
Yeah, I've finally understood. It looks like we can ditch the stray restored ancestors! Thanks for explaining!

I'll promptly work on that.

ryoqun replied:
Ok, I've successfully reverted and ci is green: https://github.com/solana-labs/solana/pull/10718/commits/4fdb41ffd500a9f2894988869a843a09d2d1ee0c

I made the error handling a hard error, considering now I'm informed that these conditions are really odd.

ryoqun commented on core/src/consensus.rs:
```diff
@@ -606,6 +703,203 @@ impl Tower {
+
+    fn stray_restored_ancestors(&self) -> HashSet<Slot> {
+        let mut ancestors: HashSet<_> = self.stray_restored_slots.clone().into_iter().collect();
+        if let Some(last_voted_slot) = self.last_voted_slot() {
+            assert!(ancestors.contains(&last_voted_slot));
+            ancestors.remove(&last_voted_slot);
+        }
```
This subtle handling is needed to be the `ancestors` of `last_voted_slot`. Also, `self.stray_restored_slots` must contain `last_voted_slot` because of `is_stray_last_vote`. (Btw, this is the reason of not being called `stray_restored_ancestors`) Otherwise, we can't construct that information...

ryoqun commented on core/src/validator.rs:
```diff
@@ -563,8 +568,71 @@ impl Validator {
+        })
+        .unwrap_or_else(|err| {
+            let is_already_active =
+                active_vote_account_exists_in_bank(&bank_forks.working_bank(), &vote_account);
```
(repost) Hmm, maybe should this be `bank_forks.root_bank()`? Otherwise, can any bad validator trick victim validator with crafted block with high-number crafted slot, somehow? I mean the victim can be tricked into the wrong perception that the vote account doesn't exist.

carllin replied:
hmmmm so if the vote account exists in the `root bank()`, then unless the vote account was cleaned up, it should also exist in the `working_bank` because the `working_bank` must be a verified/replayed descendant of the `root_bank` right?

On a side note, it's a good question what we should do if validators' saved towers do get corrupted. Without slashing, should generally be ok for correctness (no rollback) if <33% experience a failure on restart at the same time

ryoqun replied:
> hmmmm so if the vote account exists in the `root bank()`, then unless the vote account was cleaned up .... verified/replayed ...

I now understand this isn't dangerous... I misunderstood that a victim validator could tricked that its vote account is missing by being fed with crafted blocks under fully network-controlled attack scenario. Well, even such an unlikely 
dire situation, a malicious validator can't remove victim's vote account without authorization unless the private key is leaked. (implied by _verified_ in the quote)

> On a side note, it's a good question what we should do if validators' saved towers do get corrupted. 

Yeah, that's interesting consideration... Maybe renew vote account and announce its stakers to migrate as well? Or should we introduce some quarantine period (1 epoch or so) to be free of slashing? Also, should I prepare doc update for general such emergency practice guideline? (I'll do as a separate pr...) Or punt it until slashing? ;)

> Without slashing, it may be ok to just start fresh again with a new tower ...

To check my understanding, are you meaning that rollback won't occur even if sizable portion of validators' saved tower get corrupted at the same time as long as the portion is less than superminority?

carllin replied:
> we introduce some quarantine period (1 epoch or so) to be free of slashing? Also, should I prepare doc update for general such emergency practice guideline? (I'll do

Yeah hehe I think that's practically what we'll have to do, whenever we do introduce slashing. Until then, we can't really make an informed decision :)

carllin replied:
> To check my understanding, are you meaning that rollback won't occur even if sizable portion of validators' saved tower get corrupted at the same time as long as the portion is less than superminority?

Yeah, I think as long as all failures  including dead validators, tower corruption < 33%, we shouldn't see rollback

ryoqun replied:
- [ ] todo: add logging by `error` measure?

ryoqun replied:
done! 207a9c9f03d31ed4b3dad91e3f45536a601b0339

ryoqun commented:
@carllin Please review in your free time. Sorry for being rather a big PR but this is mostly same as #9902.

carllin commented on core/src/consensus.rs:
```diff
@@ -651,6 +945,144 @@ impl Tower {
+        let last_blockstore_root = blockstore.last_root();
+        if last_blockstore_root < tower_root {
+            let new_roots: Vec<_> = blockstore
+                .slot_meta_iterator(last_blockstore_root + 1)?
```
The problem with `slot_meta_iterator` is that it will iterate through any slots in blockstore in order, even if those slots are not necessarily descended from each other.

Because we only want to set roots for parents of `tower_root`, we should look up the SlotMeta for `tower_root` and traverse its parents until we reach `last_blockstore_root`, and set all those parents as roots.

I think the proper iterator here would be `AncestorIterator`, similar to here: https://github.com/solana-labs/solana/blob/master/core/src/repair_weight.rs#L408

ryoqun replied:
Hooray! Thanks for spotting this before we land this.... This bug was there since #9902... 

Done in https://github.com/solana-labs/solana/pull/10718/commits/40c7bb7674ac0461a4d2826d67927b8681fb21eb

carllin commented on core/src/consensus.rs:
```diff
@@ -606,6 +703,203 @@ impl Tower {
+        assert_eq!(slot_history.check(replayed_root_slot), Check::Found);
+        // reconcile_blockstore_roots_with_tower() should already have aligned these.
+        assert!(
+            self.root().is_none() || self.root().unwrap() <= replayed_root_slot,
```
I may be missing something, but will this `self.root().unwrap() <= replayed_root_slot` assertion fail if we jump ahead and start from a snapshot in the future?

For instance, say the chain of slots is `1, 2, 3, 4, 5`, our latest root is `2`, but then we go offline and start from a snapshot for `5` that we had not previously replayed. Then `replayed_root_slot` will be `5` at startup, but `self.root() == 2`?

ryoqun replied:
>  Then `replayed_root_slot` will be `5` at startup, but `self.root() == 2`?

Well, assertion shouldn't fail. because `self.root().unwrap() <= replayed_root_slot` could be substituted to `2 <= 5`, then `true`. So the assertion evaluates to `true` without causing `panic!`.

So, I think this assertion is fine as-is.

This asserts `replayed_root_slot` must always be greater than or equal to `self.root()` as a sanity check because we're always setting all the ancestors of `self.root()` as root at `reconcile_blockstore_roots_with_tower()` before replaying. So after replaying, `replayed_root_slot` must always proceed to `self.root()` at the very least.

carllin replied:
Heh, I misread the `<=` as `>=`, sorry!

carllin commented on core/src/consensus.rs:
```diff
@@ -606,6 +703,203 @@ impl Tower {
+
+            if anchored_slot.is_none() && check == Check::Found {
+                anchored_slot = checked_slot;
+            } else if anchored_slot.is_some() && check == Check::NotFound {
```
So `slot_history` will only contain the ancestors of the snapshot bank right? So if the snapshot is of bank `3`, and our tower has votes `0, 2, 4, 5`, like so:

```
                                         0 (saved tower root)
                                          |
                                          2 (saved tower vote)
                                      /         \
                       (snapshot root) 3          4 (saved tower vote)
                                                                 \
                                                                 5 (saved tower vote)
```

The `slot_history` of bank `3` will contain slots `2` and `0`. The `anchored_slot` will be slot `2`, but slot `5` will trigger the `NotFound`.

In this case because slot 2 has a lockout of `2^3`, we will have to wait until at least some slot greater than `2 + 2^3 = 10` to vote.

ryoqun replied:
Well, I think you're missing that this iteration goes **newer slots => older slots (reverse chronological order)**: https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR807

Also, I think I've added correct test to cover the situation you just graphed there: https://github.com/solana-labs/solana/pull/10718/files#r460669570

carllin replied:
ah yeah you're totally right, I completely missed the `rev()`. Test looks great too, thanks!

ryoqun commented:
@carllin Thanks for various review-comments! I'll address them after my first bite at #11008

carllin commented:
@ryoqun No problem! I want to do a more in depth review a little later as well, just did a first pass :)

ryoqun commented on core/src/consensus.rs:
```diff
@@ -1796,4 +2233,616 @@ pub mod test {
+
+    #[test]
+    fn test_adjust_lockouts_after_replay_not_found_slots() {
+        let mut tower = Tower::new_for_tests(10, 0.9);
+        tower.record_vote(0, Hash::default());
+        tower.record_vote(1, Hash::default());
+        tower.record_vote(2, Hash::default());
+        tower.record_vote(3, Hash::default());
+
+        let mut slot_history = SlotHistory::default();
+        slot_history.add(0);
+        slot_history.add(1);
+        slot_history.add(4);
+
+        let replayed_root_slot = 4;
+        tower = tower
+            .adjust_lockouts_after_replay(replayed_root_slot, &slot_history)
+            .unwrap();
+
+        assert_eq!(tower.voted_slots(), vec![2, 3]);
+        assert_eq!(tower.root(), Some(replayed_root_slot));
```
here is the test

carllin commented on core/src/validator.rs:
```diff
@@ -563,8 +568,71 @@ impl Validator {
+                process::exit(1);
+            }
+            let not_found = if let TowerError::IOError(io_err) = &err {
+                io_err.kind() == std::io::ErrorKind::NotFound
```
Hehe once this is checked in, we might want to communicate with validators to take extra precautions with this tower. People's disks have been failing left and right these days :)

ryoqun replied:
Yeah, this is true.

carllin commented on core/src/consensus.rs:
```diff
@@ -606,6 +703,203 @@ impl Tower {
+        let mut retain_flags_for_each_vote =
+            retain_flags_for_each_vote_in_reverse.into_iter().rev();
+
+        self.lockouts
```
nice! this was the crux of the correctness guarantee, looks great 😄

ryoqun commented:
@carllin Beware of that the commit is still immature: https://github.com/solana-labs/solana/pull/10718/commits/01b5b7a4a7443b3c4f3463bbe69966aedad5a1d9 ... I don't currently understand its full consequences yet for the changed code. I merely manged to pass the tests after rebase. maybe after #11138

ryoqun commented on local-cluster/tests/local_cluster.rs:
```diff
@@ -1390,6 +1398,140 @@ fn test_optimistic_confirmation_violation() {
+        if let Ok(slot) = validator_client.get_slot_with_commitment(CommitmentConfig::recent()) {
+            trace!("current slot: {}", slot);
+            if slot > 2 {
+                // this will be the root next time a validator starts
```
It seems that after #11138, these additional changes (2330aa89d8caeee78ce9ad5649440697ad3eb982) are needed on top of original test code in #9902.

ryoqun commented on local-cluster/tests/local_cluster.rs:
```diff
@@ -1341,6 +1341,8 @@ fn test_optimistic_confirmation_violation() {
     // dead, even though it was optimistically confirmed. The smaller validator should
     // jump over to the new fork
+    // Also, remove saved tower to intentionally make the restarted validator to violate the
+    // optimistic confirmation
```
I'm not fully sure but it seems that this test (added by #11353) doesn't get along with persistent tower... Maybe, because of testing some failure case?

carllin replied:
@ryoqun ah yeah! That's actually good this test didn't work, it means the persistent tower is serving it's purpose 😄 

What's essentially happening in this test is we are trying to simulate a vote that would an optimistic confirmation violation. So we have two validators A with 51% of the stake, B with 49%:

```
1-----2 (Vote A, B, optimistically confirmed) ---3
```

 We then kill validator `A`, mark slot `2` as dead and restart. Without persistent tower, this means validator `A` will make its own fork like so:

```
1-----2 (Vote A, B, optimistically confirmed) ---3
  |
   -----------------------4 (Validator A)
```

And validator `A` will then vote on slot 4 and b/c it has 51% of the stake, validator `B` will switch over as well and roots will be made on this new fork. This is bad because the `optimistically confirmed` slot `2` was not rooted (rolled back)

However, with this PR, `A` will not be able to vote on slot `4` because it remembers it's last vote was on slot `2`, and because `B` is still voting on `fork 2`, `A` will not be able to generate a switching proof to slot `4`. Hence this test will stall!

carllin replied:
It might actually be cool to add this as another version of this test! If we don't remove the tower, we should assert that the cluster makes no progress with `cluster.check_no_new_roots`

ryoqun replied:
Ok, I'll work on this!

ryoqun commented:
@carllin Thanks for _persistent_ review. xD I've just pushed another round of commits with rebase and review updates!
I think I've addressed all the comments you've written up to the moment.

Could you take a look at again? Also, let me know if I should split this epic pr into pieces... I know this pr held open for too long time. Sorry about this!

Anyway, I hope this persistent tower thing is nearing to the merge button..

carllin commented on core/src/consensus.rs:
```diff
@@ -406,9 +490,27 @@ impl Tower {
+
                 let switch_slot_ancestors = ancestors.get(&switch_slot).unwrap();
 
                 if switch_slot == last_voted_slot || switch_slot_ancestors.contains(&last_voted_slot) {
```
I think this check needs an extra check to account for a situation as follows:

```

1->2->3 (last_vote) -> 4 -> 5 (starting snapshot) -> 6

```
In this case, if we want to vote on `6`, even though `switch_slot_ancestors` will only include `5`,  we don't need a switching proof.

I think an  check where `|| (last_vote_slot < root && self.stray_restored_slots.is_empty())` should return, should be sufficient. We don't care about cases where `last_vote_slot > root`, because if `last_vote_slot` was an ancestor of `switch_slot`,  then `last_vote_slot` must have been replayed before `switch_slot`, and thus must exist in `BankForks` and also in `switch_slot_ancestors`.

ryoqun replied:
That's really good discussion! I'm still learning as I address your reviews.

That's great you can spot these concerns by code reviews! And I think your concern is valid. After thinking about this for a while, I think there is no need to change the code, though. In short, I believe your concern has already been addressed by the other code.

Specifically, there should be no possibility for both `self.last_voted_slot() == Some(N)` and `self.stray_restored_slots.is_empty() == true` to occur at the same time.

In other words, if and only if `self.stray_restored_slots.is_empty()`, then `self.last_voted_slot() == None` and we return `NoSwitch` from `check_switch_threshold()` by `.unwrap_or(NoSwitch)` at the very end of the `fn`'s body. Conversely, if `self.last_voted_slot() == Some(N)`, `self.stray_restored_slots.is_empty()` would always be `false` and the suggested conditional would never evaluate to `true` to enter into the special-cased `if` body anyway.

The trick is that: _The other code_ ([`do_adjust_lockouts_after_replay`](https://github.com/solana-labs/solana/pull/10718/files#r469026016)) deliberately resets `self.last_vote` if `self.lockouts.votes.is_empty()` is `true` here.

My design relationale here is that _don't let the restored tower introduce another state for tower. Minimize to violate the already-existing tower invariants. Always pretend to have been there for a while as if there was no restoration._ It's already quite stateful and super important. So, let's minimize it for controlled human's cognitive load.

That reasoning leads to this: If there were no persisted tower, there would only be 2 combinations out of 4 for the two binary-state variables: `(a) actively voting: non-empty lockouts and Some(last_vote)`, `(b) just initialized: empty lockouts and None(last_vote)`. Considering this observation, any tower state which is fully older than `root` (= no stray votes) can be regarded as being technically equivalent to the `(b)` state.

Also, normalizing fully-replayed tower to the initial/(`Default`) state of tower should be philosophically sound, forgoing to introduce _restored_ tower. I think this should be desired unless ultimately unavoidable.

The downside is that this might lead to misconception as you just were worried... For that I can extract the preceding logic into its own `fn` as a cosmetic arrangement.

To recap, once `stray_restored_slots` is assigned to `last_vote_ancestors`, it's functionally and strictly `last_vote_ancestors` in every aspect. So, there should be no leaky abstraction here, like for special-case `if`. So, the cosmetic arrangement is to indicate that intention (hopefully more clearly).

Put another way, I just tried my best to avoid to make existing conditional any longer. ;)

carllin replied:
@ryoqun that makes sense, I missed the subtlety with  `do_adjust_lockouts_after_replay` there.

Thanks for the well-reasoned response, I agree this approach with the default last_vote is much cleaner :)

ryoqun commented on core/src/consensus.rs:
```diff
@@ -620,6 +720,207 @@ impl Tower {
+            .votes
+            .retain(move |_| retain_flags_for_each_vote.next().unwrap());
+
+        if self.lockouts.votes.is_empty() {
+            info!(
+                "All restored votes were behind replayed_root_slot({}); resetting root_slot and last_vote in tower!",
+                replayed_root_slot
+            );
+            // we might not have banks for those votes so just reset.
+            // That's because the votes may well past replayed_root_slot
+            self.lockouts.root_slot = None;
+            self.last_vote = Vote::default();
```
https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR889-R897

ryoqun commented on core/src/consensus.rs:
```diff
@@ -2797,6 +2805,28 @@ pub mod test {
     }
 
+    #[test]
+    fn test_adjust_lockouts_after_replay_all_not_found_even_if_rooted() {
```
this is new test for https://github.com/solana-labs/solana/pull/10718/commits/929ca6d859b697d54753e1e4b85c04c85d7822a0#diff-d848cba681535238849e7e33ed32702bR879-R885

ryoqun commented on core/src/consensus.rs:
```diff
@@ -1842,4 +2307,706 @@ pub mod test {
+        assert_eq!(tower.root(), Some(replayed_root_slot));
+    }
+
+    #[test]
+    fn test_adjust_lockouts_after_replay_all_not_found() {
+        let mut tower = Tower::new_for_tests(10, 0.9);
+        tower.record_vote(5, Hash::default());
+        tower.record_vote(6, Hash::default());
+
+        let mut slot_history = SlotHistory::default();
+        slot_history.add(0);
+        slot_history.add(1);
+        slot_history.add(2);
+        slot_history.add(7);
+
+        let replayed_root_slot = 7;
+        tower = tower
+            .adjust_lockouts_after_replay(replayed_root_slot, &slot_history)
+            .unwrap();
+
+        assert_eq!(tower.voted_slots(), vec![5, 6]);
+        assert_eq!(tower.root(), Some(replayed_root_slot));
+    }
```
test for non-rooted tower with no common slot: https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR2787-R2806

ryoqun commented on core/src/consensus.rs:
```diff
@@ -620,6 +720,215 @@ impl Tower {
+            retain_flags_for_each_vote_in_reverse.pop();
+        }
+
+        // Check for errors if not anchored
+        // Being not anchored doesn't always equate to errors: All of votes in
+        // an unrooted (= !root_checked) warming-up vote account may legally not
+        // be anchored.
+        if anchored_slot.is_none() {
```
actual error check for the not-anchored case: https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR864-R868

ryoqun commented on core/src/consensus.rs:
```diff
@@ -418,8 +520,8 @@ impl Tower {
-                // of your last vote
-                assert!(!last_vote_ancestors.contains(&switch_slot));
+                // of your last vote unless the ancestor is restored from stray slots
+                assert!(is_restored || !last_vote_ancestors.contains(&switch_slot));
```
This is needed for test_no_optimistic_confirmation_violation_with_tower. I'll look into further why this is needed later...

ryoqun commented on core/src/consensus.rs:
```diff
@@ -459,8 +561,8 @@ impl Tower {
-                    // namely the `last_vote` itself.
-                    assert!(!last_vote_ancestors.contains(candidate_slot));
+                    // namely the `last_vote` itself, unless the ancestors are restored from stray slots.
+                    assert!(is_restored || !last_vote_ancestors.contains(candidate_slot));
```
same as https://github.com/solana-labs/solana/pull/10718/files#r472557807

ryoqun commented:
@carllin I think I've addressed the last 2 remaining review comments again. Also, I've fixed really yet another merge conflicts. (This is 12th time-ish) xD Could you take a look at it?

carllin commented on core/src/consensus.rs:
```diff
@@ -418,9 +489,21 @@ impl Tower {
+                let last_vote_ancestors =
+                    ancestors.get(&last_voted_slot).unwrap_or_else(|| {
+                        if !self.is_stray_last_vote() {
+                            panic!("no ancestors found with slot: {}", last_voted_slot);
```
hmmmm this will panic if our last vote was on some fork A, and then we start from a snapshot of another fork B right? I.e. in this example: https://github.com/solana-labs/solana/pull/10718#discussion_r474468319

If the last vote was on slot `10`, then we start from a snapshot `2`, we won't have the ancestors of `10` in `ancestors`, so this seems like it will panic.

carllin commented on core/src/consensus.rs:
```diff
@@ -418,9 +489,21 @@ impl Tower {
+                            // bank_forks doesn't have corresponding data for the stray restored last vote,
+                            // meaning severe inconsistentcy between saved tower and ledger.
+                            // (corrupted or pruned ledger, or only saved tower is moved over to new setup?)
+                            panic!("Unable to get ancestors for stray last vote {:?}", self.stray_restored_slot());
```
hmmmm this will panic if our last vote was on some fork A, and then we start from a snapshot of another fork B right? I.e. in this example: https://github.com/solana-labs/solana/pull/10718#discussion_r474468319

If the last vote was on slot `10`, then we start from a snapshot `2`, then the only keys in the `ancestor` map here will be `2` and `0`, so if you try to lookup `10` in `ancestors` here, this seems like it will panic.

ryoqun replied:
yeah, it looks like I was wrong again... I'll address.

ryoqun replied:
half done...: https://github.com/solana-labs/solana/pull/10718/commits/0c121a951f6dd5687b638d24b8ef8f56bff640a1

I'll comment later how hard to fix `test_no_optimistic_confirmation_violation_with_tower`...

ryoqun replied:
the promised detailed comment is here: https://github.com/solana-labs/solana/pull/10718#discussion_r490104064

carllin commented on core/src/consensus.rs:
```diff
@@ -418,9 +489,21 @@ impl Tower {
+                let last_vote_ancestors =
+                    ancestors.get(&last_voted_slot).unwrap_or_else(|| {
+                        if !self.is_stray_last_vote() {
+                            panic!("no ancestors found with slot: {}", last_voted_slot);
```
Double checking my assumption here 😄 

A situation like this is impossible by the time we pass the `if !self.is_stray_last_vote()` check right?

```
0 (last vote) --- 2 (snapshot)
```
Because in the above case ^ `self.last_voted_slot()` should have been empty, and we should never reach here. If this is correct, could add a comment here explaining that? Thanks!

ryoqun replied:
yeah, you're right: https://github.com/solana-labs/solana/pull/10718/commits/ada39c1c0b112d07adae6a27cc64498495a25257

Also, this prompted renewed investigation around these slot book-keeping, and I made the checked even stricter and more consistent: https://github.com/solana-labs/solana/pull/10718/commits/95ba740a1d4ff96e24272f81c7f6523a7f032cba

ryoqun commented:
@carllin well, I just noticed that when we're doing cluster restart, I think saved tower must be addressed (reverted?) as well... Maybe, `solana-ledger-tool prune-tower --slot <restart slot>` or something similar? fyi: @mvines

ryoqun commented on core/src/consensus.rs:
```diff
@@ -359,6 +429,7 @@ impl Tower {
         None
     }
 
+    // root may be forcibly set by arbitrary replay root slot
```
After this pr landed, I'll make `fn root` return `Slot` instead of `Option<Slot>`. I've made certain there must be sensible `root` for every case. And `Option<Slot>` clutters bunch of this `fn`'s callsite...

carllin commented on core/src/consensus.rs:
```diff
@@ -359,6 +429,7 @@ impl Tower {
         None
     }
 
+    // root may be forcibly set by arbitrary replay root slot
```
```suggestion
    // root may be forcibly set by arbitrary replay root slot, for example from a root after replaying a snapshot
```

ryoqun replied:
done 0bfb8227e04077e10ae5107d47569fccefaba648

carllin commented on core/src/heaviest_subtree_fork_choice.rs:
```diff
@@ -498,6 +498,44 @@ impl HeaviestSubtreeForkChoice {
+                if heaviest_slot_on_same_voted_fork.is_none() {
+                    if !tower.is_stray_last_vote() {
+                        // Unless last vote is stray, we always must be able to fetch last_voted_slot's
+                        // bast_slot, justifying to panic! here.
```
`last_slot`

ryoqun replied:
Well, this should be `best_slot`, I rewritten the comment: 0bfb8227e04077e10ae5107d47569fccefaba648

carllin commented on core/src/consensus.rs:
```diff
@@ -637,36 +733,390 @@ impl Tower {
+
+        assert_eq!(slot_history.check(replayed_root_slot), Check::Found);
+        // reconcile_blockstore_roots_with_tower() should already have aligned these.
+        assert!(self.root().is_some());
```
hmmm maybe I don't think I see anything in `reconcile_blockstore_roots_with_tower ` here: https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR1096 that guarantees `self.root().is_some()`, the restored tower seems like it could have no root if the validator never set a root before restart?

ryoqun replied:
The comment `// reconcile_blocks ...` is moved a bit to clarify what it's describing: 0bfb8227e04077e10ae5107d47569fccefaba648

Also, I've changed the handling of `root` of `Tower` to always be `Some` of `Slot`. Added some justification here: 0bfb8227e04077e10ae5107d47569fccefaba648

carllin replied:
makes sense, thanks!

carllin commented:
@ryoqun Awesome, I think this is really close, thanks for all the quality hard work!

One thing I did want to double check with you. In this PR here: https://github.com/solana-labs/solana/pull/11727/files#diff-18e1918b4b42c971ae6ae8932d39a831R843, I introduced the ability to on replaying the ledger at startup, set roots at "cluster-rooted" slots (slots rooted by > 2/3 of the network).  For example, at startup


```

3 (snapshot root) -> 4 (local/blockstore root) -> 5 -> 6 (> 2/3 cluster rooted slot, a slot rooted by) -> 7 -> 8

```

In this case the BankForks object will have root `6` after being initialized, which is higher than the local/blockstore root from before the validator restarted. 

I think this should still be safe with the changes here, because it should be equivalent to starting from a snapshot for slot `6` from the network, but I wanted to confirm with you that there's no edge cases.

ryoqun commented:
@carllin I have yet to address https://github.com/solana-labs/solana/pull/10718#issuecomment-683619724 and https://github.com/solana-labs/solana/pull/10718/files#diff-a6e7be832768ca536ac5b7450a42ee09R1479. But, I've addressed all review comments regarding new strict handling around `root` of `Tower`. I hope this commit explain things enough for your concerns: https://github.com/solana-labs/solana/pull/10718/files#diff-a6e7be832768ca536ac5b7450a42ee09R1479

ryoqun commented:
> I think this should still be safe with the changes here, because it should be equivalent to starting from a snapshot for slot `6` from the network, but I wanted to confirm with you that there's no edge cases.

Yeah, I think this is safe too. I lightly read #11727.

ryoqun commented on local-cluster/tests/local_cluster.rs:
```diff
@@ -1465,6 +1474,220 @@ fn test_optimistic_confirmation_violation() {
 
+#[test]
+#[serial]
+#[ignore]
```
Towards addressing this review comment: https://github.com/solana-labs/solana/pull/10718#discussion_r467340769

@carllin Well, I found it's very hard to make a validator violate optimistic confirmation if and only if tower is removed.
The hard part is that marking a slot as dead makes a validator to violate the optimistic conf. **even if tower is persited**. And eventually, the cluster makes new roots.

The initial situation is like this:
```
A: validator A (observer)
B: validator B (violator, will be restarted)

A & B:
0 -> 1 -> 2 -> ... -> 50 -> 51 (last vote for A and B; thus opt. confirmed)

```

Then mark 51 as dead only on B and B's new view after restart:

```
[...]  = not in bank_forks

B:
0 -> 1 -> 2 -> ... -> 50 [-> 51 (stray last vote for B)]

```

Even if we keep (optimistically-confirmed) slot 51 in towrer, due to forcibly marking 51 as dead, B's `bank_forks` doesn't have the ancestors info for slot 51. So, this falls into this clause: [`let last_vote_ancestors = ... { &empty_ancestors };`](https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR526). Then, [`!last_vote_ancestors.contains(lockout_interval_start)`](https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR604) always evaluates to `true`. So we can't exclude lockout for A correctly, which is actually  _ancestor of slot 51_. Finally, B then starts to vote creating a new fork like this:

```
B:
0 -> 1 -> 2 -> ... -> 49 -> 50 [-> 51 (stray last vote for B)]
                      |
                      +-> 56 -> 57
``` 

Ultimately, A starts to vote on the new bad fork, resulting in violated optimistic confirmation for the slot 51.

The forced dead slot 51 is the problem. But this shouldn't occur in practice. So this isn't bug per se.

But I couldn't come up with a good test scenario. Any suggestions?

carllin replied:
@ryoqun, ah I see! Does that mean we can remove this: https://github.com/solana-labs/solana/pull/10718/files#diff-a6e7be832768ca536ac5b7450a42ee09R1439-R1444 as well now?

carllin replied:
Hmm I couldn't come up with a simple example, but here's an interesting alternative idea

You have validator A + B with 31% and 36%  of the stake

0 -> 1 -> 2 -> 3 (A + B vote, optimistically confirmed)

Turn off A + B, start validator C with 33% of the stake with same ledger, but only up to slot 2. Have `C` generate some blocks like:

0 -> 1-> 2 -> 4

Then restart `A` which had 31% of the stake. With the tower, from `A`'s perspective it sees:

```
0 -> 1 -> 2 -> 3 (voted)
               |
                -> 4
```

So it should not be able to vote on `4`. Without the tower, `A` will vote on `4` b/c `C` has more stake and `B`'s vote for slot `3` is not observable (there's no descendant of `3` with the votes)

ryoqun replied:
oh this looks interesting case. Because this isn't critical part of this pr, I'll address this as a follow-up pr.

First let's land this 1700-lines pr to free me of rebasing... ;)

carllin replied:
yeah totally agree :smile:

carllin commented:
@ryoqun I think we should wait to check this in until after the mainnet upgrade so that it can bake on Tds for a while

ryoqun commented:
> @ryoqun I think we should wait to check this in until after the mainnet upgrade so that it can bake on Tds for a while

@carllin Whoa. This pr finally got LGTM-ed? Yay!!!

So, let's target v1.4? So, merging is ok? and backporting (to v1.3) is not ok?. I'm getting tired of re-re-re-re-...-rebases, maybe another one due to this #12340. ;)

carllin commented:
@ryoqun yeah lets target 1.4!

mvines commented:
> So, let's target v1.4? So, merging is ok? and backporting (to v1.3) is not ok?. I'm getting tired of re-re-re-re-...-rebases, maybe another one due to this #12340. ;)

Sorry!   Want me to wait until you merge this PR?

ryoqun commented:
> > So, let's target v1.4? So, merging is ok? and backporting (to v1.3) is not ok?. I'm getting tired of re-re-re-re-...-rebases, maybe another one due to this #12340. ;)
> 
> Sorry! Want me to wait until you merge this PR?

@mvines if it's not too painful and/or time-pressured for you to rebase, please wait. I'll merge this pr in 30-min or so after final last-minute checking.

ryoqun commented:
hmm. this isn't super fun; but it looks like even un-fsynced tower save takes around 7ms (medium):

![image](https://user-images.githubusercontent.com/117807/96208822-4cf5d100-0fa9-11eb-861c-3a60fda1310c.png)

And this directly delays that amount for any level of confirmations.

CC: @carllin

ryoqun commented:
come to think of it, I remembered we sign the saved tower.

ryoqun commented:
Weekly update on the perf and error monitoring!

![image](https://user-images.githubusercontent.com/117807/97061978-d7ef5080-15d3-11eb-9e67-1e12f8101930.png)

it seems that mean of tower save time is still high. I'd like this to be well below 1ms..


After the first upgrade with tower. 

Also, it seems that there is no tower_error/tower_warn/panic metric indicating some bug in persistent tower (checked with chronograf). (Well, #12671 isn't landed, though). Already 200+ validators experienced the upgrade:

```
$ ./target/release/solana --url http://testnet.solana.com validators
Active Stake: 16705542.793110987 SOL
Current Stake: 16493998.208943831 SOL (98.73%)
Delinquent Stake: 211544.584167159 SOL (1.27%)

Stake By Version:
1.3.11   -   0 current validators ( 0.00%),   1 delinquent validators ( 0.00%)
1.3.17   -   0 current validators ( 0.00%),   1 delinquent validators ( 0.00%)
1.4.1    - 111 current validators (26.41%),   6 delinquent validators ( 0.72%)
1.4.2    - 258 current validators (72.32%),   4 delinquent validators ( 0.36%)
unknown  -   0 current validators ( 0.00%),  69 delinquent validators ( 0.18%)

...
```
